darts/timeseries.py

```diff
@@ -3,117 +3,554 @@
 ----------
 
 `TimeSeries` is the main class in `darts`. It represents a univariate or multivariate time series.
+It can represent a stochastic time series by storing several samples (trajectories).
 """
 
 import pandas as pd
 import numpy as np
-from copy import deepcopy
+import xarray as xr
 import matplotlib.pyplot as plt
-from pandas.tseries.frequencies import to_offset
 from typing import Tuple, Optional, Callable, Any, List, Union
 from inspect import signature
+from collections import defaultdict
+from pandas.tseries.frequencies import to_offset
 
 from .logging import raise_log, raise_if_not, raise_if, get_logger
 
 logger = get_logger(__name__)
 
+# dimension names in the DataArray
+# the "time" one can be different, if it has a name in the underlying Series/DataFrame.
+DIMS = ('time', 'component', 'sample')
+
 
 class TimeSeries:
-    def __init__(self,
-                 df: pd.DataFrame,
-                 freq: Optional[str] = None,
-                 fill_missing_dates: Optional[bool] = True,
-                 dummy_index: Optional[bool] = False):
+    def __init__(self, xa: xr.DataArray):
+        """
+        Wrapper around a (well formed) DataArray. Use the static factory methods to build instances unless
+        you know what you are doing.
         """
-        A TimeSeries is an object representing a univariate or multivariate time series.
+        raise_if_not(isinstance(xa, xr.DataArray), 'Data must be provided as an xarray DataArray instance. '
+                                                   'If you need to create a TimeSeries from another type '
+                                                   '(e.g. a DataFrame), look at TimeSeries factory methods '
+                                                   '(e.g. TimeSeries.from_dataframe()).', logger)
+        raise_if_not(xa.size > 0, 'The time series array must not be empty.', logger)
+        raise_if_not(len(xa.shape) == 3, 'TimeSeries require DataArray of dimensionality 3 ({}).'.format(DIMS), logger)
+
+        # Ideally values should be np.float, otherwise certain functionalities like diff()
+        # relying on np.nan (which is a float) won't work very properly.
+        raise_if_not(np.issubdtype(xa.values.dtype, np.number), 'The time series must contain numeric values only.',
+                     logger)
+        if not np.issubdtype(xa.values.dtype, np.float):
```

Comment:
There are problems with integers because we rely on `np.nan` (which is a float) for some key pieces of the logic, so it's much simpler for the time being (and OK IMO) to require TimeSeries to contain float.